From 2da3ab1382ca8e39eb1e4428103764a81fba73d3 Mon Sep 17 00:00:00 2001
From: Rich Felker <dalias@aerifal.cx>
Date: Thu, 30 Oct 2014 20:08:40 -0400
Subject: [PATCH] fix invalid access by openat to possibly-missing variadic
 mode argument

the mode argument is only required to be present when the O_CREAT or
O_TMPFILE flag is used.
---
 src/fcntl/openat.c | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/src/fcntl/openat.c b/src/fcntl/openat.c
index 634c4bf..4faeb29 100644
--- a/src/fcntl/openat.c
+++ b/src/fcntl/openat.c
@@ -6,10 +6,14 @@
 int openat(int fd, const char *filename, int flags, ...)
 {
 	mode_t mode;
-	va_list ap;
-	va_start(ap, flags);
-	mode = va_arg(ap, mode_t);
-	va_end(ap);
+
+	if ((flags & O_CREAT) || (flags & O_TMPFILE) == O_TMPFILE) {
+		va_list ap;
+		va_start(ap, flags);
+		mode = va_arg(ap, mode_t);
+		va_end(ap);
+	}
+
 	return syscall_cp(SYS_openat, fd, filename, flags|O_LARGEFILE, mode);
 }
 
-- 
2.2.0

